File: mailblastr-python/mailblastr/automations.py
"""Automations resource — DOMAIN-FIRST: every automation belongs to one of
your sending domains (``domain`` is REQUIRED on create). Only ``Events.send``
calls with the same ``domain`` trigger it."""

from typing import Any, Dict, List, TypedDict

from . import http_client
from .http_client import build_query, paginate, path_escape as _e


class CreateParams(TypedDict, total=False):
    name: str  # required
    domain: str  # REQUIRED: the sending domain this automation belongs to
    trigger: str  # e.g. 'contact.created', 'mailblastr:schedule', 'email.opened', or a custom event name
    # {'at': ISO 8601 instant, 'timezone': IANA name} — required with the
    # 'mailblastr:schedule' trigger; not accepted on any other.
    trigger_config: Dict[str, str]
    status: str  # 'enabled' | 'disabled' (default 'disabled')
    steps: List[Dict[str, Any]]  # optional inline step graph
    connections: List[Dict[str, str]]  # optional typed edges between step keys


class UpdateParams(TypedDict, total=False):
    name: str
    status: str  # 'enabled' | 'disabled'
    domain: str  # re-point at another domain (disabled automations only)
    # Re-point the trigger (disabled automations only) — same vocabulary as
    # create. Changing it away from 'mailblastr:schedule' clears trigger_config.
    trigger: str
    trigger_key: str  # graph key the trigger node is addressed by (default 'trigger')
    # {'at', 'timezone'} — update the 'mailblastr:schedule' trigger's schedule.
    trigger_config: Dict[str, str]
    connections: List[Dict[str, str]]  # disabled automations only


class AddStepParams(TypedDict, total=False):
    # 'trigger' is NOT a step type here — the trigger is set on the automation
    # itself. Documented names ('send_email', 'wait_for_event') and internal
    # ones ('send', 'wait') are both accepted.
    type: str  # required: 'delay' | 'send_email' | 'wait_for_event' | 'condition' | 'split' | 'add_to_segment' | 'contact_update' | 'contact_delete'
    config: Dict[str, Any]
    key: str


class UpdateStepParams(TypedDict, total=False):
    type: str
    config: Dict[str, Any]
    key: str


class RunListParams(TypedDict, total=False):
    limit: int
    after: str
    before: str
    status: str  # comma-separated, e.g. 'running,failed'


class AiParams(TypedDict, total=False):
    prompt: str  # required, max 2000 characters
    template_ids: List[str]  # first 10 kept
    events: List[str]  # first 10 kept
    # {'from': step or trigger key, 'type'?: edge type, 'before'?: step key} —
    # supplying it APPENDS to the existing graph; omitting it builds a whole
    # workflow and requires the automation to have no steps yet.
    attach: Dict[str, str]


class Automations:
    """``mailblastr.Automations`` — the /automations endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams
    AddStepParams = AddStepParams
    UpdateStepParams = UpdateStepParams
    RunListParams = RunListParams
    AiParams = AiParams

    @classmethod
    def create(cls, params):
        """Create an automation (``domain`` is required). POST /automations"""
        return http_client.request("POST", "/automations", params)

    @classmethod
    def get(cls, automation_id):
        """Retrieve an automation (with steps/connections/enrollments).
        GET /automations/:id"""
        return http_client.request("GET", f"/automations/{_e(automation_id)}")

    @classmethod
    def list(cls, params=None):
        """List automations. GET /automations"""
        return http_client.request("GET", f"/automations{paginate(params)}")

    @classmethod
    def update(cls, automation_id, params):
        """Update an automation. PATCH /automations/:id"""
        return http_client.request("PATCH", f"/automations/{_e(automation_id)}", params)

    @classmethod
    def add_step(cls, automation_id, params):
        """Append a step to an automation. The automation must be DISABLED —
        editing the graph of an enabled automation is a 422.
        POST /automations/:id/steps"""
        return http_client.request("POST", f"/automations/{_e(automation_id)}/steps", params)

    @classmethod
    def update_step(cls, automation_id, step_id, params):
        """Update a step. The automation must be DISABLED.
        PATCH /automations/:id/steps/:step_id"""
        return http_client.request(
            "PATCH", f"/automations/{_e(automation_id)}/steps/{_e(step_id)}", params
        )

    @classmethod
    def delete_step(cls, automation_id, step_id):
        """Delete a step from an automation. The automation must be DISABLED.
        DELETE /automations/:id/steps/:step_id"""
        return http_client.request(
            "DELETE", f"/automations/{_e(automation_id)}/steps/{_e(step_id)}"
        )

    @classmethod
    def create_with_ai(cls, automation_id, params):
        """Build or extend an automation's steps from a natural-language
        ``prompt`` ("Create with AI"). Pass ``attach`` to append to an existing
        graph; without it the automation must have no steps yet. The automation
        must be stopped, and this route is limited to 20 requests per minute
        per account. POST /automations/:id/ai"""
        return http_client.request("POST", f"/automations/{_e(automation_id)}/ai", params)

    @classmethod
    def runs(cls, automation_id, params=None):
        """List an automation's runs, optionally filtered to a comma-separated
        ``status`` list (e.g. ``'running,failed'``).
        GET /automations/:id/runs"""
        params = params or {}
        qs = build_query(
            {
                "limit": params.get("limit"),
                "after": params.get("after"),
                "before": params.get("before"),
                "status": params.get("status"),
            }
        )
        return http_client.request("GET", f"/automations/{_e(automation_id)}/runs{qs}")

    @classmethod
    def get_run(cls, automation_id, run_id):
        """Retrieve a single automation run (with its step trace).
        GET /automations/:id/runs/:run_id"""
        return http_client.request(
            "GET", f"/automations/{_e(automation_id)}/runs/{_e(run_id)}"
        )

    @classmethod
    def stop(cls, automation_id):
        """Stop an automation — prevents new runs; in-progress runs finish.
        POST /automations/:id/stop"""
        return http_client.request("POST", f"/automations/{_e(automation_id)}/stop")

    @classmethod
    def remove(cls, automation_id):
        """Delete an automation. DELETE /automations/:id"""
        return http_client.request("DELETE", f"/automations/{_e(automation_id)}")

File: mailblastr-go/automations.go
package mailblastr

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// AutomationStep is one step of an automation's graph.
type AutomationStep struct {
	// Id is absent on the synthesized trigger step.
	Id       string         `json:"id,omitempty"`
	Key      string         `json:"key"`
	Type     string         `json:"type"`
	Position int            `json:"position,omitempty"`
	Config   map[string]any `json:"config"`
}

// AutomationConnection is a typed edge between two step keys.
type AutomationConnection struct {
	From string `json:"from"`
	To   string `json:"to"`
	Type string `json:"type"`
}

// Automation is a multi-step workflow triggered by an event.
type Automation struct {
	Object     string `json:"object"`
	Id         string `json:"id"`
	AudienceId string `json:"audience_id"`
	Name       string `json:"name"`
	Trigger    string `json:"trigger"`
	// Domain is the sending domain this automation belongs to. Only
	// Events.Send calls with the same Domain trigger it.
	Domain string `json:"domain"`
	// Status is "enabled" | "disabled".
	Status string `json:"status"`
	// TriggerConfig is set only for the "mailblastr:schedule" trigger.
	TriggerConfig *AutomationTriggerConfig `json:"trigger_config,omitempty"`
	// TriggerKey names the synthetic trigger step in Connections; defaults to
	// "trigger".
	TriggerKey string `json:"trigger_key,omitempty"`
	// Steps is omitted entirely on list responses. Its first element is always
	// the synthetic trigger step (Type "trigger", no Id, no Position).
	Steps       []AutomationStep       `json:"steps,omitempty"`
	Connections []AutomationConnection `json:"connections,omitempty"`
	// Enrollments holds enrollment counts — included only on retrieve.
	Enrollments *AutomationEnrollments `json:"enrollments,omitempty"`
	// Ai summarizes what Automations.CreateWithAi added; set only on that call.
	Ai        *AutomationAiResult `json:"ai,omitempty"`
	CreatedAt string              `json:"created_at"`
	UpdatedAt string              `json:"updated_at"`
}

// AutomationEnrollments are the counts on GET /automations/:id.
type AutomationEnrollments struct {
	Active    int `json:"active"`
	Completed int `json:"completed"`
}

// AutomationStepInput is an inline step accepted on automation create.
type AutomationStepInput struct {
	// Key optionally names the step for connections.
	Key    string         `json:"key,omitempty"`
	Type   string         `json:"type"`
	Config map[string]any `json:"config,omitempty"`
}

// AutomationConnectionInput is a typed edge accepted on create/update.
type AutomationConnectionInput struct {
	From string `json:"from"`
	To   string `json:"to"`
	Type string `json:"type,omitempty"`
}

// AutomationTriggerConfig is the config for the "mailblastr:schedule"
// trigger: the automation fires ONCE at At, enrolling every contact of its
// domain's pool. Required with that trigger; not accepted on any other.
type AutomationTriggerConfig struct {
	// At is the ISO 8601 instant the automation fires (future, at most 366
	// days ahead).
	At string `json:"at"`
	// Timezone is the IANA timezone the schedule was picked in (e.g.
	// "America/New_York").
	Timezone string `json:"timezone"`
}

// CreateAutomationRequest is the payload for POST /automations. Domain is
// REQUIRED (domain-first): only Events.Send calls with the same Domain
// trigger the automation.
type CreateAutomationRequest struct {
	Name string `json:"name"`
	// Domain is REQUIRED — the sending domain this automation belongs to
	// (e.g. "yourdomain.com" — one of your domains).
	Domain string `json:"domain"`
	// Trigger is the event that starts a run: "contact.created", the built-in
	// scheduled trigger "mailblastr:schedule" (requires TriggerConfig), an
	// engagement event ("email.opened", "email.clicked", "email.replied",
	// "email.bounced", "email.delivered"), or any custom event name you send
	// via Events.Send. Usually supplied as a Steps[0] trigger step instead.
	Trigger string `json:"trigger,omitempty"`
	// TriggerKey names the trigger in Connections; defaults to "trigger".
	TriggerKey string `json:"trigger_key,omitempty"`
	// TriggerConfig is the schedule for the "mailblastr:schedule" trigger
	// ({at, timezone}). Required with that trigger; not accepted on any other.
	TriggerConfig *AutomationTriggerConfig `json:"trigger_config,omitempty"`
	// Status is the initial status: "enabled" | "disabled" (default "disabled").
	Status string `json:"status,omitempty"`
	// Steps is an optional inline step graph.
	Steps []AutomationStepInput `json:"steps,omitempty"`
	// Connections are optional typed edges between step keys.
	Connections []AutomationConnectionInput `json:"connections,omitempty"`
}

// UpdateAutomationRequest is the payload for PATCH /automations/:id.
//
// Changing Domain, Trigger, TriggerConfig or Connections requires the
// automation to be disabled first — otherwise the API answers 422.
type UpdateAutomationRequest struct {
	// Name is at most MaxAutomationNameLength characters.
	Name string `json:"name,omitempty"`
	// Status is "enabled" | "disabled".
	Status string `json:"status,omitempty"`
	// Domain re-points the automation at another of your domains (disabled
	// automations only).
	Domain string `json:"domain,omitempty"`
	// Trigger changes the event that starts a run (disabled automations only).
	Trigger string `json:"trigger,omitempty"`
	// TriggerKey renames the trigger node referenced by Connections.
	TriggerKey string `json:"trigger_key,omitempty"`
	// TriggerConfig updates the "mailblastr:schedule" trigger's schedule
	// ({at, timezone}). Only valid on automations with that trigger.
	TriggerConfig *AutomationTriggerConfig    `json:"trigger_config,omitempty"`
	Connections   []AutomationConnectionInput `json:"connections,omitempty"`
}

// AddAutomationStepRequest appends a step to an automation. The automation
// must be disabled, and Type "trigger" is rejected — the trigger lives on the
// automation, not in its step list.
type AddAutomationStepRequest struct {
	// Type is one of AutomationStepTypes (the documented aliases
	// "send_email" and "wait_for_event" are accepted too).
	Type string `json:"type"`
	// Config may also be spread across the top level of the request body;
	// this field is the explicit form.
	Config map[string]any `json:"config,omitempty"`
	// Key names the step for Connections; defaults to the new step's id.
	Key string `json:"key,omitempty"`
}

// UpdateAutomationStepRequest edits an existing step. Every field is optional;
// the automation must be disabled.
type UpdateAutomationStepRequest struct {
	Type   string         `json:"type,omitempty"`
	Config map[string]any `json:"config,omitempty"`
	Key    string         `json:"key,omitempty"`
}

// Automation limits and vocabulary enforced by the API.
const (
	// MaxAutomationNameLength caps an automation's name.
	MaxAutomationNameLength = 255
	// MaxAutomationAiPromptLength caps CreateWithAi's prompt.
	MaxAutomationAiPromptLength = 2000
	// ScheduleTrigger is the built-in one-shot schedule trigger. It is the
	// only trigger allowed to use ReservedEventPrefix, and it requires
	// TriggerConfig.
	ScheduleTrigger = "mailblastr:schedule"
	// DefaultAutomationTrigger is used when no trigger is supplied.
	DefaultAutomationTrigger = "contact.created"
)

// AutomationStepTypes are the internal step types the API accepts. On read,
// "send" and "wait" are reported as "send_email" and "wait_for_event".
var AutomationStepTypes = []string{
	"delay", "send", "wait", "condition", "split",
	"add_to_segment", "contact_update", "contact_delete",
}

// AutomationConnectionTypes are the edge types accepted by
// AutomationConnectionInput.Type ("default" is an alias for "next").
var AutomationConnectionTypes = []string{
	"next", "default", "condition_met", "condition_not_met",
	"event_received", "timeout",
}

// ListAutomationRunsRequest lists an automation's runs with an optional status
// filter.
type ListAutomationRunsRequest struct {
	Limit  int
	After  string
	Before string
	// Status filters to runs in any of these states (e.g. "running",
	// "completed", "failed", "skipped"). Sent as a comma-separated list.
	Status []string
}

// AutomationAiAttach appends AI-generated steps to an existing graph instead
// of building a new one.
type AutomationAiAttach struct {
	// From is the trigger key or an existing step key to attach after.
	From string `json:"from"`
	// Type is one of AutomationConnectionTypes minus "next" (default
	// "default").
	Type string `json:"type,omitempty"`
	// Before, when set, must be an existing step key.
	Before string `json:"before,omitempty"`
}

// AutomationAiRequest drives "Create with AI". Without Attach the automation
// must have zero steps (workflow mode); with it the generated steps are
// appended (append mode).
type AutomationAiRequest struct {
	// Prompt is required, at most MaxAutomationAiPromptLength characters.
	Prompt string `json:"prompt"`
	// TemplateIds and Events give the model context; only the first 10 of
	// each are used.
	TemplateIds []string            `json:"template_ids,omitempty"`
	Events      []string            `json:"events,omitempty"`
	Attach      *AutomationAiAttach `json:"attach,omitempty"`
}

// AutomationAiResult summarizes an AI generation.
type AutomationAiResult struct {
	AddedSteps int `json:"added_steps"`
	// Mode is "workflow" | "append".
	Mode string `json:"mode"`
}

// DeleteAutomationStepResponse is returned by Automations.DeleteStep.
type DeleteAutomationStepResponse struct {
	Id      string `json:"id"`
	Deleted bool   `json:"deleted"`
}

// AutomationRunStep is one executed step of an automation run trace.
type AutomationRunStep struct {
	Key  string `json:"key"`
	Type string `json:"type"`
	// Status is "completed" | "failed" | "skipped".
	Status      string         `json:"status"`
	StartedAt   string         `json:"started_at"`
	CompletedAt string         `json:"completed_at"`
	Output      map[string]any `json:"output"`
	Error       string         `json:"error"`
}

// AutomationRun is one enrollment/run of an automation.
type AutomationRun struct {
	Object    string `json:"object"`
	Id        string `json:"id"`
	ContactId string `json:"contact_id"`
	// ContactEmail is the email of the contact the run is for; empty if that
	// contact was deleted.
	ContactEmail string `json:"contact_email"`
	// Status is "running" | "completed" | "failed" | "cancelled" | "skipped".
	Status      string `json:"status"`
	StartedAt   string `json:"started_at"`
	CompletedAt string `json:"completed_at"`
	CreatedAt   string `json:"created_at"`
	// AutomationId is present on GET /automations/:id/runs/:runId only.
	AutomationId string              `json:"automation_id,omitempty"`
	Steps        []AutomationRunStep `json:"steps,omitempty"`
	Error        string              `json:"error,omitempty"`
}

// AutomationsService handles the /automations endpoints.
type AutomationsService struct {
	client *Client
}

// Create creates an automation (Domain is required). POST /automations
func (s *AutomationsService) Create(params *CreateAutomationRequest) (*Automation, error) {
	return s.CreateWithContext(context.Background(), params)
}

// CreateWithContext creates an automation. POST /automations
func (s *AutomationsService) CreateWithContext(ctx context.Context, params *CreateAutomationRequest) (*Automation, error) {
	return request[Automation](ctx, s.client, http.MethodPost, "/automations", params, nil)
}

// Get retrieves an automation. GET /automations/:id
func (s *AutomationsService) Get(id string) (*Automation, error) {
	return s.GetWithContext(context.Background(), id)
}

// GetWithContext retrieves an automation. GET /automations/:id
func (s *AutomationsService) GetWithContext(ctx context.Context, id string) (*Automation, error) {
	return request[Automation](ctx, s.client, http.MethodGet, "/automations/"+esc(id), nil, nil)
}

// List lists automations. GET /automations
func (s *AutomationsService) List(params *ListParams) (*ListResponse[Automation], error) {
	return s.ListWithContext(context.Background(), params)
}

// ListWithContext lists automations. GET /automations
func (s *AutomationsService) ListWithContext(ctx context.Context, params *ListParams) (*ListResponse[Automation], error) {
	return request[ListResponse[Automation]](ctx, s.client, http.MethodGet, listPath("/automations", params), nil, nil)
}

// Update updates an automation. PATCH /automations/:id
func (s *AutomationsService) Update(id string, params *UpdateAutomationRequest) (*Automation, error) {
	return s.UpdateWithContext(context.Background(), id, params)
}

// UpdateWithContext updates an automation. PATCH /automations/:id
func (s *AutomationsService) UpdateWithContext(ctx context.Context, id string, params *UpdateAutomationRequest) (*Automation, error) {
	return request[Automation](ctx, s.client, http.MethodPatch, "/automations/"+esc(id), params, nil)
}

// AddStep appends a step to an automation; returns the created step.
// POST /automations/:id/steps
func (s *AutomationsService) AddStep(id string, params *AddAutomationStepRequest) (*AutomationStep, error) {
	return s.AddStepWithContext(context.Background(), id, params)
}

// AddStepWithContext appends a step to an automation.
func (s *AutomationsService) AddStepWithContext(ctx context.Context, id string, params *AddAutomationStepRequest) (*AutomationStep, error) {
	return request[AutomationStep](ctx, s.client, http.MethodPost, "/automations/"+esc(id)+"/steps", params, nil)
}

// UpdateStep edits an existing step; returns the updated step.
// PATCH /automations/:id/steps/:stepId
func (s *AutomationsService) UpdateStep(id, stepId string, params *UpdateAutomationStepRequest) (*AutomationStep, error) {
	return s.UpdateStepWithContext(context.Background(), id, stepId, params)
}

// UpdateStepWithContext edits an existing step.
// PATCH /automations/:id/steps/:stepId
func (s *AutomationsService) UpdateStepWithContext(ctx context.Context, id, stepId string, params *UpdateAutomationStepRequest) (*AutomationStep, error) {
	return request[AutomationStep](ctx, s.client, http.MethodPatch, "/automations/"+esc(id)+"/steps/"+esc(stepId), params, nil)
}

// DeleteStep deletes a step from an automation.
// DELETE /automations/:id/steps/:stepId
func (s *AutomationsService) DeleteStep(id, stepId string) (*DeleteAutomationStepResponse, error) {
	return s.DeleteStepWithContext(context.Background(), id, stepId)
}

// DeleteStepWithContext deletes a step from an automation.
func (s *AutomationsService) DeleteStepWithContext(ctx context.Context, id, stepId string) (*DeleteAutomationStepResponse, error) {
	return request[DeleteAutomationStepResponse](ctx, s.client, http.MethodDelete, "/automations/"+esc(id)+"/steps/"+esc(stepId), nil, nil)
}

// Runs lists an automation's runs. GET /automations/:id/runs
func (s *AutomationsService) Runs(id string, params *ListParams) (*ListResponse[AutomationRun], error) {
	return s.RunsWithContext(context.Background(), id, params)
}

// RunsWithContext lists an automation's runs. GET /automations/:id/runs
func (s *AutomationsService) RunsWithContext(ctx context.Context, id string, params *ListParams) (*ListResponse[AutomationRun], error) {
	return request[ListResponse[AutomationRun]](ctx, s.client, http.MethodGet, listPath("/automations/"+esc(id)+"/runs", params), nil, nil)
}

// RunsFiltered lists an automation's runs with an optional status filter.
// GET /automations/:id/runs
func (s *AutomationsService) RunsFiltered(id string, params *ListAutomationRunsRequest) (*ListResponse[AutomationRun], error) {
	return s.RunsFilteredWithContext(context.Background(), id, params)
}

// RunsFilteredWithContext lists an automation's runs with a status filter.
// GET /automations/:id/runs
func (s *AutomationsService) RunsFilteredWithContext(ctx context.Context, id string, params *ListAutomationRunsRequest) (*ListResponse[AutomationRun], error) {
	q := url.Values{}
	if params != nil {
		if params.Limit > 0 {
			q.Set("limit", strconv.Itoa(params.Limit))
		}
		if params.After != "" {
			q.Set("after", params.After)
		}
		if params.Before != "" {
			q.Set("before", params.Before)
		}
		if len(params.Status) > 0 {
			q.Set("status", strings.Join(params.Status, ","))
		}
	}
	path := "/automations/" + esc(id) + "/runs"
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	return request[ListResponse[AutomationRun]](ctx, s.client, http.MethodGet, path, nil, nil)
}

// GetRun retrieves a single automation run. GET /automations/:id/runs/:runId
func (s *AutomationsService) GetRun(id, runId string) (*AutomationRun, error) {
	return s.GetRunWithContext(context.Background(), id, runId)
}

// GetRunWithContext retrieves a single automation run.
func (s *AutomationsService) GetRunWithContext(ctx context.Context, id, runId string) (*AutomationRun, error) {
	return request[AutomationRun](ctx, s.client, http.MethodGet, "/automations/"+esc(id)+"/runs/"+esc(runId), nil, nil)
}

// Stop stops an automation — prevents new runs; in-progress runs finish.
// POST /automations/:id/stop
func (s *AutomationsService) Stop(id string) (*Automation, error) {
	return s.StopWithContext(context.Background(), id)
}

// StopWithContext stops an automation. POST /automations/:id/stop
func (s *AutomationsService) StopWithContext(ctx context.Context, id string) (*Automation, error) {
	return request[Automation](ctx, s.client, http.MethodPost, "/automations/"+esc(id)+"/stop", nil, nil)
}

// CreateWithAi generates automation steps from a natural-language prompt and
// merges them into the automation, which must be stopped (disabled). The
// result's Ai field reports how many steps were added.
//
// Rate limited to 20 requests per 60 s per account (rate_limit_exceeded, 429),
// and it spends AI credits (ai_credits_exceeded, 429, when exhausted).
// POST /automations/:id/ai
func (s *AutomationsService) CreateWithAi(id string, params *AutomationAiRequest) (*Automation, error) {
	return s.CreateWithAiWithContext(context.Background(), id, params)
}

// CreateWithAiWithContext generates automation steps from a prompt.
// POST /automations/:id/ai
func (s *AutomationsService) CreateWithAiWithContext(ctx context.Context, id string, params *AutomationAiRequest) (*Automation, error) {
	return request[Automation](ctx, s.client, http.MethodPost, "/automations/"+esc(id)+"/ai", params, nil)
}

// Remove deletes an automation. DELETE /automations/:id
func (s *AutomationsService) Remove(id string) (*RemovedResponse, error) {
	return s.RemoveWithContext(context.Background(), id)
}

// RemoveWithContext deletes an automation. DELETE /automations/:id
func (s *AutomationsService) RemoveWithContext(ctx context.Context, id string) (*RemovedResponse, error) {
	return request[RemovedResponse](ctx, s.client, http.MethodDelete, "/automations/"+esc(id), nil, nil)
}
